pkg/services/sqlstore/migrations/dashboard_mig.go

```diff
@@ -234,4 +234,10 @@ func addDashboardMigration(mg *Migrator) {
 	mg.AddMigration("Add isPublic for dashboard", NewAddColumnMigration(dashboardV2, &Column{
 		Name: "is_public", Type: DB_Bool, Nullable: false, Default: "0",
 	}))
+
+	mg.AddMigration("clean up mangled slugs", NewRawSQLMigration("").
+		SQLite("UPDATE dashboard SET slug=replace(replace(replace(replace(slug,'%2f',''),'%',''),'---','-'),'--','-') WHERE slug GLOB '*[-%]*'").
```

Comment:
This makes me nervous :)

Reply:
Me too, open to alternative suggestions...

Reply:
We could consider handling this outside of core Grafana -  add a temporary "dashboard slug migration"-specific CLI command, for example, and tell users who are running into these issues to run that against their (backed up) database. Users who aren't seeing issues never have to run it, and we call UpdateSlug whenever a dashboard is saved (and - I think - updated), so in theory all used slugs will get updated eventually anyway. 

We can limit how long we keep the command around to, say, just this major version of grafana - if you start with Grafana 11, you never need to know about this; if you're upgrading from v9, upgrade to v10.latest first and run that command (eventually that command could run a full suite of helpful db things: "check your tables and schemas and is that even a database we support, etc" along with "data migrations that probably shouldn't run when grafana starts and instead require manual intervention") 

----

A second option is more simply to check for `%` in slugs during Get and re-write them at runtime. We can't necessarily update the database at the same time, but we'll know they get updated in the database eventually (unless the dash in question never gets updated). It's not the most satisfying option but it's also not really different than other runtime data validation and transformation-type tasks (imo).

Reply:
Something outside Grafana unfortunately won't help on cloud unless we also run it at which point we'd likely be adding more risk than simply fixing things.  My comment was somewhat tongue in cheek since database updates like this do always have some level of risk, but realistically the sql is relatively straightforward and the risk of changing a slug is quite low at this point because they are mostly there to make the urls look nice rather than to actually address the dashboard.

Changing the slug on the fly at runtime has roughly the same set of risks, but we'd never be able to remove it due to the possibility of users updating from a "bad" release directly to the current release.  It also doesn't have the safety valve of being able to edit the dashboard and change the slug knowing it won't be changed again after the migration has been run.

That said, I do think we need to look at the SQL with a critical eye and think about any possible cases where it would fail, and I'm open to simplifying it by only modifying the % encoding and not worrying about the duplicated dashes since they are not causing an actual user problem.